LanguageModeling/GPT/oneflow_gpt/distribute.py

```diff
@@ -0,0 +1,303 @@
+        node_ids = [i // args.num_gpus_per_node for i in range(self.world_size_)]
+        device_ids = list(range(args.num_gpus_per_node)) * args.num_nodes
+        node_device_ids = [f"{n}:{d}" for n, d in zip(node_ids, device_ids)]
+        return node_device_ids
+
+    def _init_parallel_hierarchy(self):
+        if self.is_hybrid_parallel():
+            self.parallel_hierarchy_ = [
+                self.data_parallel_size_,
+                self.tensor_model_parallel_size_,
+            ]
+        else:
+            self.parallel_hierarchy_ = None
+
+    @property
+    def parallel_hierarchy(self):
+        return self.parallel_hierarchy_
+
+    @property
+    def tensor_model_parallel_size(self):
+        return self.tensor_model_parallel_size_
+
+    def get_layer_placement(self, layer_idx):
+        return self.layers_device_ids_[layer_idx]
+
+    def get_layer_stage(self, layer_idx):
+        return self.layers_stage_ids_[layer_idx]
+
+    def is_model_parallel(self):
+        return self.tensor_model_parallel_size_ > 1
+
+    def is_data_parallel(self):
+        return self.data_parallel_size_ > 1
+
+    def is_pipeline_parallel(self):
+        return self.pipeline_model_parallel_size_ > 1
+
+    def is_hybrid_parallel(self):
+        return self.is_model_parallel() and self.is_data_parallel()
+
+    def is_non_parallel(self):
+        return not self.is_model_parallel() and not self.is_data_parallel()
+
+
+def get_dist_util():
+    global _DIST_UTIL
+    if _DIST_UTIL is None:
+        _DIST_UTIL = _DistributeUtil()
+    return _DIST_UTIL
+
+
+@contextlib.contextmanager
+def layer_placement_scope(layer_idx, device="gpu"):
+    dist_util = get_dist_util()
+    with flow.scope.placement(
+        device, dist_util.get_layer_placement(layer_idx), dist_util.parallel_hierarchy,
+    ):
+        if dist_util.is_pipeline_parallel():
+            with flow.experimental.scope.config(
+                pipeline_stage_id_hint=dist_util.get_layer_stage(layer_idx)
+            ):
+                yield
+        else:
+            yield
+
+
+@contextlib.contextmanager
+def data_placement_scope():
+    with layer_placement_scope(0, "cpu"):
+        yield
+
+
+def _is_sbp_str(s):
+    if not isinstance(s, str):
+        return False
+
+    if s in ("B", "P"):
+        return True
+
+    if s.startswith("S(") and s[-1] == ")":
+        return True
+
+    return False
+
+
+def _gen_parallel_dist_by_2d_sbp(sbp_list):
+    assert isinstance(sbp_list, list)
+    assert len(sbp_list) == 2
+    assert all(_is_sbp_str(sbp) for sbp in sbp_list)
+
+    dist_util = get_dist_util()
+    if dist_util.is_hybrid_parallel():
+        return sbp_list
+    elif dist_util.is_data_parallel():
+        return sbp_list[:1]
+    elif dist_util.is_model_parallel():
+        return sbp_list[1:]
+    elif dist_util.is_non_parallel():
+        return []
+    else:
+        raise NotImplementedError
+
+
+def get_data_parallel_dist():
```

Comment:
let's discuss this function and following functions again.
All of them are called only once in model.py. It will more easy for users to understand SBP transfer on site if `_gen_parallel_dist_by_2d_sbp(["S(0)", "B"])` is called but not `get_data_parallel_dist`.  These functions are not necessary at all.

Suggest to change `_gen_parallel_dist_by_2d_sbp` name and use it on site.

Reply:
不是很赞同，我希望 spb 和 parallel_cast 相关的处理全部集中放在 distribute.py 里面，model.py 里面 focus 计算图的结构。可以讨论。